packages/deployment/lib/kes.js

```diff
@@ -148,44 +148,81 @@ class UpdatedKes extends Kes {
     }
   }
 
+  /**
+   * build CloudWatch alarm widgets
+   *
+   * @param {string[]} alarmNames list of alarm names
+   * @param {Object} alarmTemplate widget template for alarm
+   * @returns {Object[]} list of alarm widgets
+   */
+  buildAlarmWidgets(alarmNames, alarmTemplate) {
+    return alarmNames.map((alarmName) => {
+      const alarm = cloneDeep(alarmTemplate);
+      alarm.properties.title = alarmName;
+      alarm.properties.annotations.alarms[0] = alarm.properties.annotations.alarms[0].replace('alarmTemplate', alarmName);
+      return alarm;
+    });
+  }
+
   /**
    * build CloudWatch dashboard based on the dashboard configuration and other configurations
    *
    * @param {Object} dashboardConfig dashboard configuration for creating widgets
-   * @param {Object} es elastic search configuration including configuration for alarms
+   * @param {Object} ecs Elastic Container Service configuration including custom configuration
+   * for alarms
+   * @param {Object} es Elasticsearch configuration including configuration for alarms
    * @param {string} stackName stack name
    * @returns {string} returns dashboard body string
    */
-  buildCWDashboard(dashboardConfig, es, stackName) {
-    const esTemplateAlarm = dashboardConfig.esTemplateAlarm;
+  buildCWDashboard(dashboardConfig, ecs, es, stackName) {
+    const alarmTemplate = dashboardConfig.alarmTemplate;
+
+    // build ECS alarm widgets
+    const ecsAlarmNames = [];
+    Object.keys(ecs.services).forEach((serviceName) => {
+      // default alarm
+      const defaultAlarmName = `${stackName}-${serviceName}-TaskCountLowAlarm`;
+      ecsAlarmNames.push(defaultAlarmName);
+      // custom alarm
+      if (ecs.services[serviceName].alarms) {
+        Object.keys(ecs.services[serviceName].alarms).forEach((alarmName) => {
+          const name = `${stackName}-${serviceName}-${alarmName}Alarm`;
+          ecsAlarmNames.push(name);
+        });
+      }
+    });
+
+    const ecsAlarms = this.buildAlarmWidgets(ecsAlarmNames, alarmTemplate);
 
     // build ES alarm widgets
-    const alarms = Object.keys(es.alarms).map((alarmName) => {
-      const esAlarm = cloneDeep(esTemplateAlarm);
-      const title = `${stackName}-${es.name}-${alarmName}Alarm`;
-      esAlarm.properties.title = title;
-      esAlarm.properties.annotations.alarms[0] = esAlarm.properties.annotations.alarms[0].replace('esTemplateAlarm', title);
-      return esAlarm;
-    });
+    let esWidgets = [];
+    if (es) {
+      const esAlarmNames = Object.keys(es.alarms).map((alarmName) =>
+        `${stackName}-${es.name}-${alarmName}Alarm`);
+      const esAlarms = this.buildAlarmWidgets(esAlarmNames, alarmTemplate);
+      esWidgets = dashboardConfig.esHeader
+        .concat(cloneDeep(dashboardConfig.alarmHeader), esAlarms, dashboardConfig.esWidgets);
```

Comment:
What is the cloneDeep here to protect against?

Reply:
dashboardConfig.alarmHeader can be added to the dashboard multiple places.  widgets are updated later with x, y values, so it's necessary to have separate copy.